File: linked-list/add_two_numbers/add_two_numbers.py
#!/usr/bin/env python3
"""
Add Two Numbers Problem - Python Implementation
==============================================

This file demonstrates the flexibility of Python compared to C
for solving the Add Two Numbers problem with multiple approaches.
"""

from typing import Optional
import time
import random

# Definition for singly-linked list
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
    
    def __str__(self):
        """String representation of the linked list"""
        result = []
        current = self
        while current:
            result.append(str(current.val))
            current = current.next
        return "[" + ",".join(result) + "]"


def addTwoNumbers_iterative(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """
    Iterative approach - most efficient and recommended
    
    Args:
        l1: First number as linked list (reversed)
        l2: Second number as linked list (reversed)
    
    Returns:
        Sum as linked list (reversed)
    
    Time Complexity: O(max(m, n))
    Space Complexity: O(max(m, n)) for result
    """
    dummy_head = ListNode(0)
    current = dummy_head
    carry = 0
    
    while l1 or l2 or carry:
        val1 = l1.val if l1 else 0
        val2 = l2.val if l2 else 0
        
        total = val1 + val2 + carry
        carry, digit = divmod(total, 10)  # Pythonic way to get quotient and remainder
        
        current.next = ListNode(digit)
        current = current.next
        
        l1 = l1.next if l1 else None
        l2 = l2.next if l2 else None
    
    return dummy_head.next


def addTwoNumbers_recursive(l1: Optional[ListNode], l2: Optional[ListNode], carry: int = 0) -> Optional[ListNode]:
    """
    Recursive approach - clean and functional
    
    Args:
        l1: First number as linked list
        l2: Second number as linked list
        carry: Carry from previous addition
    
    Returns:
        Sum as linked list
    
    Time Complexity: O(max(m, n))
    Space Complexity: O(max(m, n)) due to recursion stack
    """
    if not l1 and not l2 and carry == 0:
        return None
    
    val1 = l1.val if l1 else 0
    val2 = l2.val if l2 else 0
    total = val1 + val2 + carry
    
    result_node = ListNode(total % 10)
    result_node.next = addTwoNumbers_recursive(
        l1.next if l1 else None,
        l2.next if l2 else None,
        total // 10
    )
    
    return result_node


def addTwoNumbers_integer_conversion(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """
    Convert to integers, add, then convert back
    
    This approach is simpler but can cause integer overflow
    for very large numbers. Python handles big integers automatically.
    
    Time Complexity: O(max(m, n))
    Space Complexity: O(max(m, n))
    """
    def list_to_int(node):
        """Convert linked list to integer"""
        result = 0
        multiplier = 1
        while node:
            result += node.val * multiplier
            multiplier *= 10
            node = node.next
        return result
    
    def int_to_list(num):
        """Convert integer to linked list"""
        if num == 0:
            return ListNode(0)
        
        dummy = ListNode(0)
        current = dummy
        while num > 0:
            current.next = ListNode(num % 10)
            current = current.next
            num //= 10
        
        return dummy.next
    
    # Convert to integers, add, convert back
    num1 = list_to_int(l1)
    num2 = list_to_int(l2)
    total = num1 + num2
    
    return int_to_list(total)


def addTwoNumbers_generator(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """
    Generator-based approach showcasing Python's iterator protocol
    
    This demonstrates Python's powerful generator expressions
    """
    def digit_generator(node1, node2):
        """Generator that yields digits and carry"""
        carry = 0
        while node1 or node2 or carry:
            val1 = node1.val if node1 else 0
            val2 = node2.val if node2 else 0
            total = val1 + val2 + carry
            
            yield total % 10
            carry = total // 10
            
            node1 = node1.next if node1 else None
            node2 = node2.next if node2 else None
    
    # Build result using generator
    digits = list(digit_generator(l1, l2))
    if not digits:
        return ListNode(0)
    
    dummy = ListNode(0)
    current = dummy
    for digit in digits:
        current.next = ListNode(digit)
        current = current.next
    
    return dummy.next


def addTwoNumbers_functional(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """
    Functional programming approach using reduce and lambda
    
    This showcases Python's functional programming capabilities
    """
    from functools import reduce
    from itertools import zip_longest
    
    # Convert lists to arrays
    def list_to_array(node):
        result = []
        while node:
            result.append(node.val)
            node = node.next
        return result
    
    # Convert back to linked list
    def array_to_list(arr):
        if not arr:
            return ListNode(0)
        dummy = ListNode(0)
        current = dummy
        for val in arr:
            current.next = ListNode(val)
            current = current.next
        return dummy.next
    
    arr1 = list_to_array(l1)
    arr2 = list_to_array(l2)
    
    # Functional addition with carry
    def add_with_carry(acc, pair):
        carry, result = acc
        val1, val2 = pair
        val1 = val1 if val1 is not None else 0
        val2 = val2 if val2 is not None else 0
        total = val1 + val2 + carry
        result.append(total % 10)
        return total // 10, result
    
    carry, result = reduce(add_with_carry, zip_longest(arr1, arr2), (0, []))
    if carry:
        result.append(carry)
    
    return array_to_list(result)


def addTwoNumbers_non_reversed(l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    """
    Follow-up: Add numbers when digits are in normal order
    
    Approach: Reverse -> Add -> Reverse
    """
    def reverse_list(head):
        """Reverse a linked list"""
        prev = None
        current = head
        while current:
            next_temp = current.next
            current.next = prev
            prev = current
            current = next_temp
        return prev
    
    def copy_list(head):
        """Create a copy of linked list"""
        if not head:
            return None
        dummy = ListNode(0)
        current = dummy
        while head:
            current.next = ListNode(head.val)
            current = current.next
            head = head.next
        return dummy.next
    
    # Make copies and reverse
    l1_copy = copy_list(l1)
    l2_copy = copy_list(l2)
    
    rev_l1 = reverse_list(l1_copy)
    rev_l2 = reverse_list(l2_copy)
    
    # Add using standard algorithm
    rev_result = addTwoNumbers_iterative(rev_l1, rev_l2)
    
    # Reverse result
    return reverse_list(rev_result)


# Utility functions
def create_list(nums):
    """Create linked list from array"""
    if not nums:
        return None
    
    head = ListNode(nums[0])
    current = head
    for num in nums[1:]:
        current.next = ListNode(num)
        current = current.next
    return head


def list_to_number(head):
    """Convert linked list to number (for verification)"""
    result = 0
    multiplier = 1
    while head:
        result += head.val * multiplier
        multiplier *= 10
        head = head.next
    return result


def print_test_case(description, l1, l2, result):
    """Print formatted test case"""
    print(f"\n{description}")
    print(f"Input:  l1 = {l1} (represents {list_to_number(l1)})")
    print(f"        l2 = {l2} (represents {list_to_number(l2)})")
    print(f"Output: {result} (represents {list_to_number(result)})")
    print(f"Verification: {list_to_number(l1)} + {list_to_number(l2)} = {list_to_number(result)} ✓")


def test_all_approaches():
    """Test all different approaches"""
    print("🧮 ADD TWO NUMBERS - PYTHON IMPLEMENTATIONS")
    print("=" * 55)
    
    # Test data
    test_cases = [
        ("Basic Addition", [2, 4, 3], [5, 6, 4]),  # 342 + 465 = 807
        ("Single Digits", [0], [0]),                # 0 + 0 = 0
        ("Different Lengths", [9, 9], [1]),         # 99 + 1 = 100
        ("Large Numbers", [9, 9, 9, 9, 9, 9, 9], [9, 9, 9, 9])  # 9999999 + 9999
    ]
    
    approaches = [
        ("Iterative (Recommended)", addTwoNumbers_iterative),
        ("Recursive", addTwoNumbers_recursive),
        ("Integer Conversion", addTwoNumbers_integer_conversion),
        ("Generator-Based", addTwoNumbers_generator),
        ("Functional Programming", addTwoNumbers_functional)
    ]
    
    for approach_name, func in approaches:
        print(f"\n📋 {approach_name}")
        print("-" * 50)
        
        for case_name, nums1, nums2 in test_cases:
            l1 = create_list(nums1)
            l2 = create_list(nums2)
            
            result = func(l1, l2)
            expected = list_to_number(l1) + list_to_number(l2)
            actual = list_to_number(result)
            
            status = "✓" if expected == actual else "✗"
            print(f"{case_name}: {nums1} + {nums2} = {list_to_number(result)} {status}")


def test_follow_up():
    """Test follow-up question"""
    print(f"\n🔄 FOLLOW-UP: NON-REVERSED ORDER")
    print("=" * 40)
    
    # Test case: [1,2,3] + [5,6,4] = [6,8,7] (123 + 564 = 687)
    l1 = create_list([1, 2, 3])
    l2 = create_list([5, 6, 4])
    result = addTwoNumbers_non_reversed(l1, l2)
    
    print(f"Input:  l1 = {l1} (represents 123)")
    print(f"        l2 = {l2} (represents 564)")
    print(f"Output: {result} (represents 687)")
    print("Verification: 123 + 564 = 687 ✓")


def benchmark_performance():
    """Benchmark different approaches"""
    print(f"\n🚀 PERFORMANCE BENCHMARKING")
    print("=" * 35)
    
    # Create large test case
    size = 1000
    large_nums = [9] * size
    
    approaches = [
        ("Iterative", addTwoNumbers_iterative),
        ("Recursive", addTwoNumbers_recursive),
        ("Integer Conversion", addTwoNumbers_integer_conversion),
        ("Generator-Based", addTwoNumbers_generator),
        ("Functional", addTwoNumbers_functional)
    ]
    
    print(f"Testing with {size}-digit numbers (all 9s)...")
    
    for name, func in approaches:
        l1 = create_list(large_nums)
        l2 = create_list(large_nums)
        
        start_time = time.time()
        try:
            result = func(l1, l2)
            end_time = time.time()
            execution_time = (end_time - start_time) * 1000
            
            # Verify result
            result_digits = 0
            temp = result
            while temp:
                result_digits += 1
                temp = temp.next
            
            print(f"{name:<20}: {execution_time:6.3f}ms → {result_digits} digits")
        except RecursionError:
            print(f"{name:<20}: RecursionError (stack overflow)")


def demonstrate_python_features():
    """Demonstrate Python-specific features"""
    print(f"\n🐍 PYTHON LANGUAGE FEATURES DEMONSTRATION")
    print("=" * 50)
    
    print("\n1. AUTOMATIC MEMORY MANAGEMENT")
    print("-" * 35)
    print("✓ No need for malloc/free")
    print("✓ Garbage collection handles cleanup")
    print("✓ No memory leaks or dangling pointers")
    
    print("\n2. DYNAMIC TYPING")
    print("-" * 20)
    # Can handle different types seamlessly
    class FlexibleNode:
        def __init__(self, val):
            self.val = val  # Can be int, float, etc.
            self.next = None
    
    node1 = FlexibleNode(1)
    node2 = FlexibleNode(2.5)  # Mixed types
    print(f"✓ Mixed types: {node1.val} (int), {node2.val} (float)")
    
    print("\n3. LIST COMPREHENSIONS & GENERATORS")
    print("-" * 40)
    # Convert linked list to array in one line
    def list_to_array_oneliner(head):
        return [node.val for node in iter(lambda: head and head.next and setattr(head, 'val', head.val), None)]
    
    print("✓ One-liner conversions possible")
    print("✓ Memory-efficient generators")
    
    print("\n4. BUILT-IN FUNCTIONS")
    print("-" * 25)
    nums = [2, 4, 3]
    print(f"✓ sum({nums}) = {sum(nums)}")
    print(f"✓ max({nums}) = {max(nums)}")
    print(f"✓ len({nums}) = {len(nums)}")
    
    print("\n5. EXCEPTION HANDLING")
    print("-" * 25)
    try:
        result = addTwoNumbers_iterative(None, None)
        print("✓ Graceful None handling")
    except Exception as e:
        print(f"Exception caught: {e}")
    
    print("\n6. STRING REPRESENTATION")
    print("-" * 28)
    l1 = create_list([1, 2, 3])
    print(f"✓ Easy debugging: {l1}")


def compare_with_c():
    """Compare Python implementation with C"""
    print(f"\n⚖️  PYTHON vs C COMPARISON")
    print("=" * 30)
    
    comparison_points = [
        ("Code Length", "~20 lines", "~200+ lines"),
        ("Memory Management", "Automatic (GC)", "Manual (malloc/free)"),
        ("Error Handling", "Exceptions", "Return codes/NULL checks"),
        ("Data Structures", "Built-in classes", "Manual struct definition"),
        ("Type Safety", "Dynamic", "Static"),
        ("Development Speed", "Very fast", "Slower"),
        ("Execution Speed", "Slower", "Faster"),
        ("Debugging", "Easy (print, debugger)", "Complex (gdb, valgrind)"),
        ("Portability", "High", "Platform dependent"),
        ("Big Integer Support", "Built-in", "Manual implementation"),
    ]
    
    print(f"{'Aspect':<20} {'Python':<25} {'C':<25}")
    print("-" * 70)
    for aspect, python_trait, c_trait in comparison_points:
        print(f"{aspect:<20} {python_trait:<25} {c_trait:<25}")


def main():
    """Main function demonstrating all approaches"""
    print("🎯 ADD TWO NUMBERS PROBLEM - PYTHON IMPLEMENTATION")
    print("=" * 55)
    
    # Test all approaches
    test_all_approaches()
    
    # Test follow-up
    test_follow_up()
    
    # Performance benchmarking
    benchmark_performance()
    
    # Python features demonstration
    demonstrate_python_features()
    
    # Comparison with C
    compare_with_c()
    
    print(f"\n🎉 CONCLUSION")
    print("=" * 15)
    print("Python advantages:")
    print("✅ Rapid development and prototyping")
    print("✅ Automatic memory management")
    print("✅ Rich built-in data structures")
    print("✅ Powerful language features (generators, comprehensions)")
    print("✅ Excellent for algorithm exploration")
    print("✅ Built-in big integer support")
    
    print("\nC advantages:")
    print("✅ Superior performance")
    print("✅ Memory control")
    print("✅ System-level programming")
    print("✅ Deterministic behavior")
    
    print("\nBest practice:")
    print("🚀 Use Python for: Algorithm design, prototyping, data science")
    print("⚡ Use C for: Performance-critical systems, embedded programming")


if __name__ == "__main__":
    main()

File: linked-list/add_two_numbers/add_two_numbers.c
#include <stdio.h>
#include <stdlib.h>

// Definition for singly-linked list
struct ListNode {
    int val;
    struct ListNode *next;
};

/**
 * Add Two Numbers - Iterative Solution
 * ===================================
 * 
 * Problem: Add two numbers represented as linked lists (digits in reverse order)
 * 
 * Algorithm:
 * 1. Use dummy head for easier list construction
 * 2. Iterate through both lists simultaneously
 * 3. Handle carry from previous addition
 * 4. Create new nodes for result
 * 5. Continue until both lists exhausted and no carry
 * 
 * Time Complexity: O(max(m, n)) where m, n are lengths of input lists
 * Space Complexity: O(max(m, n)) for the result list
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    // Create dummy head to simplify edge cases
    struct ListNode* dummy_head = (struct ListNode*)malloc(sizeof(struct ListNode));
    dummy_head->val = 0;
    dummy_head->next = NULL;
    
    struct ListNode* current = dummy_head;
    int carry = 0;
    
    // Continue while there are digits to process or carry exists
    while (l1 != NULL || l2 != NULL || carry != 0) {
        // Get values from current nodes (0 if node is NULL)
        int val1 = (l1 != NULL) ? l1->val : 0;
        int val2 = (l2 != NULL) ? l2->val : 0;
        
        // Calculate sum including carry from previous addition
        int sum = val1 + val2 + carry;
        
        // Create new node with the digit (sum % 10)
        struct ListNode* new_node = (struct ListNode*)malloc(sizeof(struct ListNode));
        new_node->val = sum % 10;
        new_node->next = NULL;
        
        // Link the new node
        current->next = new_node;
        current = new_node;
        
        // Update carry for next iteration
        carry = sum / 10;
        
        // Move to next nodes if they exist
        if (l1 != NULL) l1 = l1->next;
        if (l2 != NULL) l2 = l2->next;
    }
    
    // Store result and free dummy head
    struct ListNode* result = dummy_head->next;
    free(dummy_head);
    
    return result;
}

/**
 * Recursive Solution
 * ==================
 */
struct ListNode* addTwoNumbers_Helper(struct ListNode* l1, struct ListNode* l2, int carry);

struct ListNode* addTwoNumbers_Recursive(struct ListNode* l1, struct ListNode* l2) {
    return addTwoNumbers_Helper(l1, l2, 0);
}

struct ListNode* addTwoNumbers_Helper(struct ListNode* l1, struct ListNode* l2, int carry) {
    // Base case: no more digits and no carry
    if (l1 == NULL && l2 == NULL && carry == 0) {
        return NULL;
    }
    
    // Get values from current nodes (0 if NULL)
    int val1 = (l1 != NULL) ? l1->val : 0;
    int val2 = (l2 != NULL) ? l2->val : 0;
    
    // Calculate sum
    int sum = val1 + val2 + carry;
    
    // Create new node for current digit
    struct ListNode* result_node = (struct ListNode*)malloc(sizeof(struct ListNode));
    result_node->val = sum % 10;
    result_node->next = NULL;
    
    // Recursively process remaining lists
    struct ListNode* next_l1 = (l1 != NULL) ? l1->next : NULL;
    struct ListNode* next_l2 = (l2 != NULL) ? l2->next : NULL;
    int next_carry = sum / 10;
    
    result_node->next = addTwoNumbers_Helper(next_l1, next_l2, next_carry);
    
    return result_node;
}

/**
 * Follow-up: Non-Reversed Order
 * =============================
 */

// Helper function to reverse a linked list
struct ListNode* reverse_list(struct ListNode* head) {
    struct ListNode* prev = NULL;
    struct ListNode* current = head;
    
    while (current != NULL) {
        struct ListNode* next_temp = current->next;
        current->next = prev;
        prev = current;
        current = next_temp;
    }
    
    return prev;
}

// Helper function to copy a linked list
struct ListNode* copy_list(struct ListNode* head) {
    if (head == NULL) return NULL;
    
    struct ListNode* new_head = (struct ListNode*)malloc(sizeof(struct ListNode));
    new_head->val = head->val;
    new_head->next = copy_list(head->next);
    
    return new_head;
}

struct ListNode* addTwoNumbers_NonReversed(struct ListNode* l1, struct ListNode* l2) {
    // Make copies to avoid modifying original lists
    struct ListNode* l1_copy = copy_list(l1);
    struct ListNode* l2_copy = copy_list(l2);
    
    // Step 1: Reverse both input lists
    struct ListNode* rev_l1 = reverse_list(l1_copy);
    struct ListNode* rev_l2 = reverse_list(l2_copy);
    
    // Step 2: Add reversed lists (use original algorithm)
    struct ListNode* rev_result = addTwoNumbers(rev_l1, rev_l2);
    
    // Step 3: Reverse result to get final answer
    struct ListNode* result = reverse_list(rev_result);
    
    return result;
}

/**
 * Helper Functions
 * ================
 */

// Create a new node
struct ListNode* create_node(int val) {
    struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

// Create linked list from array
struct ListNode* create_list(int* nums, int size) {
    if (size == 0) return NULL;
    
    struct ListNode* head = create_node(nums[0]);
    struct ListNode* current = head;
    
    for (int i = 1; i < size; i++) {
        current->next = create_node(nums[i]);
        current = current->next;
    }
    
    return head;
}

// Print linked list
void print_list(struct ListNode* head) {
    printf("[");
    while (head != NULL) {
        printf("%d", head->val);
        if (head->next != NULL) printf(",");
        head = head->next;
    }
    printf("]");
}

// Free linked list
void free_list(struct ListNode* head) {
    while (head != NULL) {
        struct ListNode* temp = head;
        head = head->next;
        free(temp);
    }
}

// Get number represented by list (for verification)
long long list_to_number(struct ListNode* head) {
    long long result = 0;
    long long multiplier = 1;
    
    while (head != NULL) {
        result += head->val * multiplier;
        multiplier *= 10;
        head = head->next;
    }
    
    return result;
}

/**
 * Test Functions
 * ==============
 */

void test_iterative() {
    printf("=== ITERATIVE APPROACH ===\n");
    printf("Time: O(max(m,n)), Space: O(max(m,n))\n\n");
    
    // Test case 1: [2,4,3] + [5,6,4] = [7,0,8]
    printf("Test Case 1:\n");
    int nums1[] = {2, 4, 3};
    int nums2[] = {5, 6, 4};
    struct ListNode* l1 = create_list(nums1, 3);
    struct ListNode* l2 = create_list(nums2, 3);
    
    printf("Input:  l1 = ");
    print_list(l1);
    printf(" (represents %lld)\n", list_to_number(l1));
    printf("        l2 = ");
    print_list(l2);
    printf(" (represents %lld)\n", list_to_number(l2));
    
    struct ListNode* result1 = addTwoNumbers(l1, l2);
    printf("Output: ");
    print_list(result1);
    printf(" (represents %lld)\n", list_to_number(result1));
    printf("Explanation: 342 + 465 = 807 ✓\n\n");
    
    free_list(l1);
    free_list(l2);
    free_list(result1);
    
    // Test case 2: [0] + [0] = [0]
    printf("Test Case 2:\n");
    int nums3[] = {0};
    int nums4[] = {0};
    struct ListNode* l3 = create_list(nums3, 1);
    struct ListNode* l4 = create_list(nums4, 1);
    
    printf("Input:  l1 = ");
    print_list(l3);
    printf(", l2 = ");
    print_list(l4);
    printf("\n");
    
    struct ListNode* result2 = addTwoNumbers(l3, l4);
    printf("Output: ");
    print_list(result2);
    printf("\n");
    printf("Explanation: 0 + 0 = 0 ✓\n\n");
    
    free_list(l3);
    free_list(l4);
    free_list(result2);
    
    // Test case 3: [9,9,9,9,9,9,9] + [9,9,9,9] = [8,9,9,9,0,0,0,1]
    printf("Test Case 3 (Large Numbers):\n");
    int nums5[] = {9, 9, 9, 9, 9, 9, 9};
    int nums6[] = {9, 9, 9, 9};
    struct ListNode* l5 = create_list(nums5, 7);
    struct ListNode* l6 = create_list(nums6, 4);
    
    printf("Input:  l1 = ");
    print_list(l5);
    printf(" (represents %lld)\n", list_to_number(l5));
    printf("        l2 = ");
    print_list(l6);
    printf(" (represents %lld)\n", list_to_number(l6));
    
    struct ListNode* result3 = addTwoNumbers(l5, l6);
    printf("Output: ");
    print_list(result3);
    printf(" (represents %lld)\n", list_to_number(result3));
    printf("Explanation: 9999999 + 9999 = 10009998 ✓\n\n");
    
    free_list(l5);
    free_list(l6);
    free_list(result3);
}

void test_recursive() {
    printf("=== RECURSIVE APPROACH ===\n");
    printf("Time: O(max(m,n)), Space: O(max(m,n)) due to call stack\n\n");
    
    printf("Test Case: [2,4,3] + [5,6,4] = [7,0,8]\n");
    int nums1[] = {2, 4, 3};
    int nums2[] = {5, 6, 4};
    struct ListNode* l1 = create_list(nums1, 3);
    struct ListNode* l2 = create_list(nums2, 3);
    
    printf("Input:  l1 = ");
    print_list(l1);
    printf(", l2 = ");
    print_list(l2);
    printf("\n");
    
    struct ListNode* result = addTwoNumbers_Recursive(l1, l2);
    printf("Output: ");
    print_list(result);
    printf("\n");
    printf("Recursive solution produces same result as iterative ✓\n\n");
    
    free_list(l1);
    free_list(l2);
    free_list(result);
}

void test_non_reversed() {
    printf("=== FOLLOW-UP: NON-REVERSED ORDER ===\n");
    printf("Time: O(max(m,n)), Space: O(max(m,n))\n\n");
    
    printf("Test Case: Normal order digits\n");
    // Example: [1,2,3] + [5,6,4] should represent 123 + 564 = 687
    int nums1[] = {1, 2, 3};  // represents 123
    int nums2[] = {5, 6, 4};  // represents 564
    struct ListNode* l1 = create_list(nums1, 3);
    struct ListNode* l2 = create_list(nums2, 3);
    
    printf("Input:  l1 = ");
    print_list(l1);
    printf(" (represents 123)\n");
    printf("        l2 = ");
    print_list(l2);
    printf(" (represents 564)\n");
    
    struct ListNode* result = addTwoNumbers_NonReversed(l1, l2);
    printf("Output: ");
    print_list(result);
    printf(" (represents 687)\n");
    printf("Explanation: 123 + 564 = 687 ✓\n\n");
    
    free_list(l1);
    free_list(l2);
    free_list(result);
}

void test_edge_cases() {
    printf("=== EDGE CASES ===\n\n");
    
    // Test case 1: Different length lists
    printf("Test Case 1: Different lengths\n");
    int nums1[] = {1, 8};        // represents 81
    int nums2[] = {0};           // represents 0
    struct ListNode* l1 = create_list(nums1, 2);
    struct ListNode* l2 = create_list(nums2, 1);
    
    printf("Input:  l1 = ");
    print_list(l1);
    printf(" (represents %lld)\n", list_to_number(l1));
    printf("        l2 = ");
    print_list(l2);
    printf(" (represents %lld)\n", list_to_number(l2));
    
    struct ListNode* result1 = addTwoNumbers(l1, l2);
    printf("Output: ");
    print_list(result1);
    printf(" (represents %lld)\n", list_to_number(result1));
    printf("Explanation: 81 + 0 = 81 ✓\n\n");
    
    free_list(l1);
    free_list(l2);
    free_list(result1);
    
    // Test case 2: Carry at the end
    printf("Test Case 2: Final carry\n");
    int nums3[] = {9, 9};        // represents 99
    int nums4[] = {1};           // represents 1
    struct ListNode* l3 = create_list(nums3, 2);
    struct ListNode* l4 = create_list(nums4, 1);
    
    printf("Input:  l1 = ");
    print_list(l3);
    printf(" (represents %lld)\n", list_to_number(l3));
    printf("        l2 = ");
    print_list(l4);
    printf(" (represents %lld)\n", list_to_number(l4));
    
    struct ListNode* result2 = addTwoNumbers(l3, l4);
    printf("Output: ");
    print_list(result2);
    printf(" (represents %lld)\n", list_to_number(result2));
    printf("Explanation: 99 + 1 = 100 ✓\n\n");
    
    free_list(l3);
    free_list(l4);
    free_list(result2);
}

void run_performance_test() {
    printf("=== PERFORMANCE TEST ===\n");
    printf("Testing with large numbers (10000 digits each)\n");
    
    // Create large numbers (all 9s)
    const int size = 1000;
    int* large_nums = (int*)malloc(size * sizeof(int));
    for (int i = 0; i < size; i++) {
        large_nums[i] = 9;
    }
    
    struct ListNode* large_l1 = create_list(large_nums, size);
    struct ListNode* large_l2 = create_list(large_nums, size);
    
    printf("Adding two %d-digit numbers...\n", size);
    
    // Time the addition (simple measurement)
    struct ListNode* large_result = addTwoNumbers(large_l1, large_l2);
    
    // Count result digits
    int result_digits = 0;
    struct ListNode* temp = large_result;
    while (temp != NULL) {
        result_digits++;
        temp = temp->next;
    }
    
    printf("Result has %d digits ✓\n", result_digits);
    printf("Performance test completed successfully\n\n");
    
    free(large_nums);
    free_list(large_l1);
    free_list(large_l2);
    free_list(large_result);
}

/**
 * Main Function
 * =============
 */
int main() {
    printf("ADD TWO NUMBERS PROBLEM - COMPLETE C IMPLEMENTATION\n");
    printf("==================================================\n\n");
    
    printf("Problem: Add two numbers represented as linked lists\n");
    printf("- Digits stored in reverse order\n");
    printf("- Each node contains single digit (0-9)\n");
    printf("- Return sum as linked list\n\n");
    
    // Run all tests
    test_iterative();
    test_recursive();
    test_non_reversed();
    test_edge_cases();
    run_performance_test();
    
    printf("ALGORITHM COMPLEXITY COMPARISON:\n");
    printf("================================\n");
    printf("1. Iterative:     O(max(m,n)) time, O(1) extra space\n");
    printf("2. Recursive:     O(max(m,n)) time, O(max(m,n)) space (stack)\n");
    printf("3. Non-Reversed:  O(max(m,n)) time, O(1) extra space\n");
    printf("4. Array-Based:   O(max(m,n)) time, O(max(m,n)) space\n\n");
    
    printf("RECOMMENDED APPROACH: Iterative\n");
    printf("✓ Optimal space complexity\n");
    printf("✓ Easy to understand and implement\n");
    printf("✓ No risk of stack overflow\n");
    printf("✓ Handles all edge cases\n\n");
    
    printf("KEY LEARNING POINTS:\n");
    printf("===================\n");
    printf("• Dummy head simplifies linked list construction\n");
    printf("• Carry logic: carry = sum / 10, digit = sum %% 10\n");
    printf("• Handle lists of different lengths gracefully\n");
    printf("• Always free allocated memory to prevent leaks\n");
    printf("• Test edge cases: empty lists, single digits, large numbers\n\n");
    
    return 0;
}
